### DISCUSSION TITLE
Roslyn Analyzers for FluentMigrator

### DISCUSSION THREAD

**Discussion by jzabroski** [fluentmigrator/fluentmigrator#1390] [Ideas]
1. `AsDecimal(const x, const y)` validates x precision is not > 28 (reason: .NET does not allow decimal values greater than 28 digits).
2. `Alter.Table("Foo").InSchema("dbo").AddColumn("x").AsDecimal()`
    - Analyzer should warn that you should specify `NotNullable`
    - Fixer that adds NotNullable to all code in the file where this happens.
3. `Alter.Table("Foo").InSchema("dbo").AddColumn("x").AsDateTime();`
    - Should warn that you should use `AsDateTime2` instead, as it is ANSI DateTime.
    - Fixer should automatically replace all `AsDateTime` with `AsDateTime2`.
4. `Alter.Table("Foo").InSchema("dbo").AddColumn("x").AsDateTime2();`
    - Should warn that you didn't specify the precision
    - Should make sure this is easy to disable, as I personally would find this annoying.  The other reason is different databases currently have different max precision value.
5. `Alter.Table("foo").InSchema("dbo").AddColumn("x").AsDecimal().NotNullable();` should warn that you need to use `SetExistingRowsTo` to avoid run-time errors where the column is not seeded.  See discussion in #1379 - it doesn't need to be `AsDecimal` and can be for any data type.
6. Existing PR #1240 from @benmccallum to support analyzer ideas covered in #1184
    - Analyzer that checks for duplicate migration numbers
    - Fixer that checks if there is a gap to move the duplicate "up" or "down" one version number.
7. `Delete.FromTable("Foo").InSchema("dbo");`
    - Analyzer should warn that you should specify `.Row();` or `.AllRows();`
    - Fixer 1 (default behavior) should add `.Row();`
    - Fixer 2 should add `.AllRows();`
8. Analyze whether certain extensions require `TransactionBehavior.None`. Ex: #1385 Postgres Concurrently
    - Analyzer 1 should warn if `[Migration(int, string, TransactionBehavior.Default]`
    - Fixer 1 for Analyzer 1 (default behavior) should convert it to `[Migration(int, string, TransactionBehavior.None]`
    - Analyzer 2 should warn if TransactionBehavior is not specified, e.g.,  `[Migration(int, string)]`
    - Fixer 1 for Analyzer 2 (default behavior) should convert it to `[Migration(int, string, TransactionBehavior.None]`
    - Should we have a `string TransactionBehaviorJustification` overload to the MigrationAttribute constructor?  Common justifications might be "knowable" by the analyzer, so that if you remove Concurrently, it realizes the justification is no longer valid?  Too fancy/complex?
9. Incompatible Options for Postgres Index Options
10. Duplicate column name in Create Table expression, e.g.
    ```c#
    Alter.Table("Foo").InSchema("dbo")
      .WithColumn("CustomNote1").AsString().Nullable()
      .WithColumn("CustomNote1").AsString().Nullable();
    ```
    - Analyzer should warn that you have a duplicate column name CustomNote1.
    - Fixer 1 (default behavior) should replace duplicate with Same Bears emoji. ʕ ⁰̈ᴥ⁰̈ʔʕ⁰̈ᴥ⁰̈ ʔ 💯 
11. Migration sub-class has a `[Tags("SomeTag")]` attribute but no `[Migration(0)]` attribute
    - Fixer 1 (default behavior) generate migration attribute with highest available migration id (Based on source control)

**Comment by jzabroski** [comment]
@fubar-coder I was just bit by writing `Delete.FromTable("Foo").InSchema("dbo");` without a WHERE clause.  Curiously, the announcer says:

```
Beginning Transaction 
December 14th 2020 16:16:03Info
FluentMigrator.Expressions.DeleteDataExpression 
December 14th 2020 16:16:03Info
FluentMigrator.Expressions.DeleteDataExpression 
December 14th 2020 16:16:03Info
Committing Transaction 
```

**Reply by fubar-coder** [reply]
I don't know why the announcer shows two delete expressions, but this seems like a bug. The ability to execute a `DELETE` without a `WHERE` is usually a bug, but be aware that changing this will be a breaking change.

**Comment by PhenX** [comment]
I started to work on analyzers 1 and 2 here https://github.com/PhenX/fluentmigrator/pull/2

 * AsDecimal precision via tha Range attribute, 
 * NotNullable/Nullable requirement

**Reply by jzabroski** [reply]
Does that include the Fixer? I added one comment here https://github.com/PhenX/fluentmigrator/pull/2/commits/9ecf9e9a2a59fafa5a7c48cae7914a367ee57ab8#r2021833853

**Reply by PhenX** [reply]
Nope, not yet, I never wrote a fixer for now, that'll be fun

**Reply by jzabroski** [reply]
I would study how NUnit does it. I was really impressed at how easy they made my life upgrading FluentMigrator to the latest API. I've personally preferred xUnit.net for the longest time, even though FluentMigrator uses NUnit. I have to say, I encounter slightly fewer bugs with NUnit and zero analyzer/fixer bugs, whereas XUnit has issues when you do things like put a common ABC  TestBase in a nuget package and that TestBase has a [Fact] in it that your concrete tests inherit, etc.

**Reply by PhenX** [reply]
I'm a little more used to Xunit, but was surprised but the features I integrated in the last pull requests, it works great!
About the code fixer : I made it, but I'm not sure if it should suggest NotNullable() or Nullable() (or both ?), because columns are nullable if not specified am I right ?

**Reply by jzabroski** [reply]
NotNullable is the default but it surprises some people, perhaps especially for string data. If not too much trouble, offering both makes sense but the default suggestion should continue to be NotNullable.

**Comment by ArthurSilvaAutoLab** [comment]
I want to suggest an Analyzer:

Some Oracle versions (between 10 and 19c, I believe) don't allow names bigger than 30 characters. It would be awesome if an Analyzer could check it and warns the 30+ names. Does it make sense? I don't know if this kind of feature fits here...

Inspired by @PhenX PR I will try to work on it, and see how it goes. It's a painful problem for us...

**Reply by PhenX** [reply]
Yep, the infamous "ORA-00972: IDENTIFIER IS TOO LONG", of Oracle 12.1 and below, while with Oracle 12.2+ it's 128, it bit us hard many times too.
Any identifier is concerned : table name, column name, constraint name, index name, etc

Even generated names of course, so it may not always be caught by an analyzer, sadly. Of course it could ofetn help, but could give a false sense of safety, maybe it should be coupled with a runtime check.

There is already a [check for constraint name](https://github.com/fluentmigrator/fluentmigrator/blob/main/src/FluentMigrator.Runner.Oracle/Generators/Oracle/OracleColumn.cs#L29) but that's far from enough.

**Reply by ArthurSilvaAutoLab** [reply]
Exactly this one!
I agree with the perspective of this being coupled with the runtime check. And we understand this is not going to be bulletproof, but to us seems more and more that an false sense of safety is better than no sense at all 🤣 

Would you suggest to extend this into the generator?

**Reply by PhenX** [reply]
An analyzer would be a good start ! 
Feel free to propose it, I think it should check inline identifiers, or constants like that : 

```csharp
Create.Table("foo");

const string tableName;
Create.Table(tableName);
```

And include table, column, constraint, and index.
